### ISSUE TITLE
Apple support is complaining about Location Services permission text

### ISSUE THREAD

**Issue by rolinger** [dpa99c/cordova-diagnostic-plugin#290]
Ok...so this is a new one.  I submitted my most recent version of my app to Apple and their support team rejected it with the message of:

> "we noticed that your app requests the user’s consent to access their Location, but does not appropriately provide the main reason for the use of this feature in the permission modal alert.
> 
> To resolve this issue, please revise the permission modal alert to specify why the app is requesting access to the user's Location."
> 

The message that is appearing during permission request is:

```
"Allow 'THISAPP' to access your location while you are using this app?  
This app requires access to your location when the screen is on and the app is displayed"

```
Last I checked we don't have the ability to modify the Permissions Request verbiage - we can't change if we wanted to, its some that is controlled by iOS itself.  I am not supplying that message from my code, and as best I can tell the `diagnostics` plugin does not create that message either - thus further demonstrating its something we can't control.  I am arguing this point with Apple for like 2 days now and they just keep saying the same thing.

Am I missing something...does this plugin allow us to customize the permission modal alert text?

**Comment by rolinger** [comment]
Apple Support truly does suck.  I would say 50% of the time I submit a new version of my app its rejected from some new BS reason.  My most recent rejection was for two reasons.  1.  My app failed to initialize and load on iPad.  But this has been the case for all 19 prior versions...why suddenly now is this an issue with Apple.  My app was built for iPhone only and they are now telling me it must render on iPad as well.  The 2nd issue is defined above - but that same "location" verbiage has been the same for all previous 19 versions as well - or what ever the default message was for each iOS rev.  So why is it a problem now.

I really hate Apple.

**Comment by dpa99c** [comment]
This plugin contains [default iOS usage descriptions for location permission](https://github.com/dpa99c/cordova-diagnostic-plugin/blob/master/plugin.xml#L50) since if they are omitted entirely, the binary will be rejected on upload (which means it can't even be used in TestFlight).

But the expectation is that the using app will customise these messages by overriding the plugin defaults in the `config.xml`. Since `cordova@7` parsing of `<config-file>` blocks in `config.xml` is supported without needing any additional scripts/plugins, so you can just place the blocks in your `config.xml` and they will override the values in this plugin's `plugin.xml`, e.g.

	<config-file target="*-Info.plist" parent="NSLocationAlwaysAndWhenInUseUsageDescription">
	    <string>This app requires constant access to your location in order to spy on you</string>
	</config-file>

**Comment by dpa99c** [comment]
With regard to Apple, I agree entirely. It's Russian roulette as to which reviewer you get. Some will let almost anything past and some pick up on the most trivial details. There's no consistency. The same binary can get passed by one reviewer and rejected by another. Or the same reviewer, depending if they're having a bad day...

**Comment by rolinger** [comment]
HAHAHAHA....just saw the "...in order to spy on you" part.   Obviously I had no idea about this NSLocation...Despription setting.  I looked it up and it said since iOS 10 - man, I have had 15 versions of my app on iOS 10, no one ever mentioned this issue before.  And quite honestly, this guy isn't happy with iOS own default message.  

One wouldn't think iOS (or Android for that matter) would allow users to customize permission messages - it really gives some one a chance to deceive or lie to users about the real intent.  It's also counter-intuitive for Apple too...considering their strict and absolute control over everything, why would they give developers the chance to change default messaging?

**Comment by dpa99c** [comment]
The NSUsageDescription's are seen by Apple as how you justify to your users that your need the permission being asked for. Since the message is displayed to the user and they have the options to "Allow" or "Don't Allow", the theory is that the user can make an informed choice. In reality, if they "Don't Allow", 90% of apps will not work properly. 

It's worth noting that some iOS permission messages (such as requesting permission for notifications) are hard-coded and can't be customised via a UsageDescription. My strategy to manage these has been a "pre-permission-dialog dialog", i.e. "I'm about to ask your permission to send you Push Notifications. iOS is going to then show you a really stupid message, but please ignore it and press "Allow" otherwise you won't be notified of cool shit".

**Comment by rolinger** [comment]
LOL...thanks for the help and directing me on this one.  Thought it was related to how `diagnostics` does permission request , turns out its geolocation specific setting...and a grumpy iOS reviewer.

**Comment by rolinger** [comment]
It turns out the cordova-diagnostic-plugin was responsible for the NSLocation messaging.  These iOS Location Description settings were not found in the geolocation plugin (where I thought I would find them).  Instead I found multiple cordova-diagnostic-plugin references to NSLocation..Descriptions in the following files:

\appProject\node_modules\cordova.plugins.diagnostic\plugin.xml
\appProject\plugins\cordova.plugins.diagnostic\plugin.xml

I guess they were added in iOS 11 but I only deployed my first app for iOS 11 in like November so I didn't realize these things had been added.  For future I am just adding `<config-file target="*-Info.plist"...>` for the following three settings:

NSLocationAlwaysUsageDescription
NSLocationWhenInUseUsageDescription  // this one seems to be the key
NSLocationAlwaysAndWhenInUseUsageDescription

Even when I did the recommendation above for `NSLocationAlwaysAndWhenInUseUsageDescription` the  one built into the plugin (NSLocationWhenInUseUsageDescription)  was overriding my custom one.   So I just changed all three to be the same custom message.

On a side note, I went through 3 different messages with Apple before they finally approved one.  Jackholes they are!
